src/pathfinder.rs

```diff
@@ -0,0 +1,283 @@
+        sqrt(
+            ((self.x as isize - other.x as isize).pow(2)
+                + (self.y as isize - other.y as isize).pow(2)) as usize,
+        )
+    }
+}
+
+#[derive(Error, Debug)]
+enum RegisteringNodesError {
+    #[error(transparent)]
+    Serialization(#[from] serde_json::Error),
+    #[error("Nodes were not correctly indexed")]
+    NodesNotCorrectlyIndexed,
+}
+
+byond_fn!(fn register_nodes_astar(json) {
+    match register_nodes(json) { Ok(s) => Some(s),
+        Err(e) => Some(format!("{e}"))
+    }
+});
+
+// Builds a list of nodes from a json file.
+// Errors if the input list of nodes is not correctly indexed. Each node should have for unique id its position in the list, with the first unique-id being 0.
+fn register_nodes(json: &str) -> Result<String, RegisteringNodesError> {
+    let deserialized_nodes: Vec<Node> = serde_json::from_str(json)?;
+    if deserialized_nodes
+        .iter()
+        .enumerate()
+        .filter(|(i, node)| i != &node.unique_id)
+        .count()
+        != 0
+    {
+        return Err(RegisteringNodesError::NodesNotCorrectlyIndexed);
+    }
+
+    deserialized_nodes
+        .into_iter()
+        .for_each(|node| push_node(node));
+
+    Ok("1".to_string())
+}
+
+byond_fn!(fn add_node_astar(json) {
+    match add_node(json) {
+        Ok(s) => Some(s),
+        Err(e) => Some(format!("{e}"))
+    }
+});
+
+// Add a node to the static list of node.
+// If it is connected to other existing nodes, it will update their connected_nodes_id list.
+fn add_node(json: &str) -> Result<String, RegisteringNodesError> {
+    let new_node: Node = serde_json::from_str(json)?;
+
+    // As always, a node unique id should correspond to its index in NODES
+    if new_node.unique_id != get_nodes_len() {
+        return Err(RegisteringNodesError::NodesNotCorrectlyIndexed);
+    }
+
+    // Make sure every connection we have we other nodes is 2 ways
+    new_node.connected_nodes_id.iter().for_each(|index| {
+        NODES.with(|nodes_ref| {
+            if let Some(Some(node)) = nodes_ref.borrow_mut().get_mut(*index) {
+                Rc::get_mut(node)
+                    .unwrap()
+                    .connected_nodes_id
+                    .push(new_node.unique_id)
+            }
+        })
+    });
+
+    push_node(new_node);
+
+    Ok("1".to_string())
+}
+
+#[derive(Error, Debug)]
+enum DeleteNodeError {
+    #[error(transparent)]
+    ParsingError(ParseIntError),
+    #[error("Node was not found")]
+    NodeNotFound,
+}
+
+byond_fn!(fn remove_node_astar(unique_id) {
+    match remove_node(unique_id) {
+        Ok(s) => Some(s),
+        Err(e) => Some(format!("{e}"))
+    }
+});
+
+// Replace the node with unique_id by None
+// Update connected nodes as well so nothing target the removed node anymore
+// Errors if no node can be found with unique_id
+fn remove_node(unique_id: &str) -> Result<String, DeleteNodeError> {
+    let unique_id = match unique_id.parse::<usize>() {
```

Comment:
I failed to use ? (error propagation) with the thiserror crate. Did I miss anything or is it impossible ?